### DISCUSSION TITLE
PEM Format

### DISCUSSION THREAD

**Discussion by luciano-souza** [lsnepomuceno/laravel-a1-pdf-sign#147] [General]
Does it work with a .pem format certificate?

**Comment by lsnepomuceno** [comment]
Hello @luciano-souza,

It would be necessary to convert to crt, which is completely viable using OpenSSL.

I will try to simulate the conversion with a self-signed certificate.

But if you have a certificate that it would be okay to share just to validate this conversion, I would be happy to help, just send me to lsn.nepomuceno@gmail.com.

I will get back to you as soon as I have any news regarding the tests.

**Comment by lsnepomuceno** [comment]
Hello @luciano-souza,

Yes — as of **2.1.0**, released today. The conversion to `.crt` / PKCS#12 I suggested back then is no longer necessary: a PEM certificate is handed to the signer directly.

**Certificate and private key in the same file:**

```php
use LSNepomuceno\LaravelA1PdfSign\Facades\A1PdfSign;

$signed = A1PdfSign::signFromPem('path/to/certificate.pem', 'password', 'path/to/document.pdf');

$signed->save('path/to/signed.pdf');
```

**Private key in a file of its own:**

```php
$signed = A1PdfSign::newSignature()
    ->certificatePem('path/to/certificate.crt', 'path/to/private.key', 'password')
    ->pdf('path/to/document.pdf')
    ->sign();
```

**From bytes you already hold** — an upload, a secret manager, a database column:

```php
$signed = A1PdfSign::newSignature()
    ->certificateFromPem($certificateBytes, $keyBytes)
    ->pdf('path/to/document.pdf')
    ->sign();
```

Three things worth knowing:

- **The extension does not matter.** PEM ships as `.pem`, `.crt`, `.cer`, `.key` and `.txt`, so the encoding is read from the file's content. Nothing gates on the suffix.
- **The password may be empty** when the private key is unencrypted — legal for PEM, impossible for PKCS#12. Prefer an encrypted key where you have the choice: an unprotected one is readable by anything that can read the file.
- **Everything after the certificate is unchanged.** PAdES profiles, timestamps, visible seals, multiple signatures and validation all behave identically, because PKCS#12 and PEM converge on one pipeline as soon as the certificate is parsed.

The `pdf:sign` command reads PEM too, with `--key` for the two-file form:

```bash
php artisan pdf:sign document.pdf certificate.crt '' --key=private.key
```

Docs: [Working with certificates](https://laravel-a1-pdf-sign.netlify.app/docs/2.x/working-with-certificate) · [Sign PDF file](https://laravel-a1-pdf-sign.netlify.app/docs/2.x/sign-pdf-file) · [Release notes](https://github.com/lsnepomuceno/laravel-a1-pdf-sign/releases/tag/2.1.0)

Note that 2.1.0 requires **PHP 8.4+ and Laravel 13** — if you are still on the 1.x line, that upgrade comes first ([guide](https://github.com/lsnepomuceno/laravel-a1-pdf-sign/blob/main/UPGRADE.md)).

Sorry it took this long. Thanks for raising it — please open an issue if anything misbehaves with a real certificate.
